docs/proposals/20210310-opt-in-autoscaling-from-zero.md

```diff
@@ -0,0 +1,242 @@
+The cluster autoscaler will then utilize this information by reading the appropriate infrastructure
+reference from the resource it is scaling (MachineSet or MachineDeployment).
+
+A user may override the field in the associated infrastructure template by applying annotations to the
+MachineSet or MachineDeployment in question. In these cases the autoscaler will evaluate the annotation
+information in favor of reading the information from the status.
+
+### User Stories
+
+#### Story 1
+
+As a Cluster API user, I would like to reduce my operating costs by scaling down my workload
+cluster when they are not in use. Using the cluster autoscaler with a minimum size of zero for
+a MachineSet or MachineDeployment will allow me to automate the scale down actions for my clusters.
+
+#### Story 2
+
+As a Cluster API user, I would like to have special resource nodes (eg GPU enabled) provided when needed by workloads
+without the need for human intervention. As these nodes might be more expensive, I would also like to return them when
+not in use. By using the cluster autoscaler with a zero-sized MachineSet or MachineDeployment, I can automate the
+creation of nodes that will not consume resources until they are required by applications on my cluster.
+
+### Implementation Details/Notes/Constraints
+
+Cloud providers, or users, can add an optional field to their machine template infrastructure objects to indicate
+the resources provided by the machine. Internally, this field will be represented by a Go `map` type
+utilizing named constants for the keys and `k8s.io/apimachinery/pkg/api/resource.Quantity` as the values
+(similar to how resource limits and requests are handled for pods).
+
+**Example implementation in Docker provider**
+```
+// these constants will be carried in Cluster API, but are repeated here for clarity
+const (
+    AutoscalerResourceCPU ResourceName = "cpu"
+    AutoscalerResourceMemory ResourceName = "memory"
+    AutoscalerResourceGPU ResourceName = "gpu"
+)
+
+// DockerMachineTemplateSpec defines the desired state of DockerMachineTemplate.
+type DockerMachineTemplateSpec struct {
+    AutoscalerCapacity ResourceList `json:"autoscalerCapacity,omitempty"`
+    Template DockerMachineTemplateResource `json:"template"`
+}
+```
+_Note: the `ResourceList` and `ResourceName` referenced are from k8s.io/kubernetes/pkg/apis/core/types`_
+
+When used as a manifest, it would look like this:
+
+```
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+kind: DockerMachineTemplate
+metadata:
+  name: workload-md-0
+  namespace: default
+spec:
+  autoscalerCapacity:
+    memory: 500mb
+    cpu: "1"
+    gpu: "1"
+  template:
+    spec: {}
+```
+
+If a user wishes to override the resources hints as provided from the machine template, they may
+do so by adding the following annotations to their MachineSets or MachineDeployments:
+
+```
+kind: MachineSet
```

Comment:
TBH I have some concern about the entire idea of overrides.
- Afaik templates de facto are limited to a cluster/to a machineDeployment due to how we define ownerRef and how cascading delete works, so the infrastructure machine is already specific to a set of machines. 
- If we want to allow overrides on one of the CAPI resources, should those go on the MachineDeployment instead, given that MachineSet are kind of an internal detail e they comes and go automatically, thus forcing the user to re-apply override every time a new MS gets created
- Having two places where the same information can be set could be kind of confusing

What about unblocking the scale for zero problem and then iterate on UX as soon as we have more concrete feedback from the field?

Reply:
>     * Afaik templates de facto are limited to a cluster/to a machineDeployment due to how we define ownerRef and how cascading delete works, so the infrastructure machine is already specific to a set of machines.

perhaps i misunderstood something then, is it possible to have a single infrastructure machine template that is linked in 2 separate machineDeployments?

the other thing about this is that the autoscaler resource requests in the infrastructure machine are optional, so the overrides would provide an extremely lightweight way to add scale from zero to a single group of machines without having to change the infrastructure.

>     * If we want to allow overrides on one of the CAPI resources, should those go on the MachineDeployment instead, given that MachineSet are kind of an internal detail e they comes and go automatically, thus forcing the user to re-apply override every time a new MS gets created

my thought here is that allowing it on both resources makes it easy to use for multiple styles of user.

>     * Having two places where the same information can be set could be kind of confusing

yeah, i think that's fair. we would document with examples, but there is no escaping that this would be the same information in multiple places (if utilized).

> What about unblocking the scale for zero problem and then iterate on UX as soon as we have more concrete feedback from the field?

i could be convinced to drop this ;)

but, on the ux front, these annotations are how we use scale from zero currently with machinesets on openshift. i think it would be cool if we contribute that part of feature as well. i appreciate the feedback Fabrizio!

Reply:
I think it's possible to add it to several MachineDeployments but not to multiple clusters. The InfrastructureMachineTemplate has a ownerReferences to the cluster, i.e. when the cluster is deleted the InfrastructureMachineTemplate is also deleted.